common/filesystem.cpp

```diff
@@ -0,0 +1,515 @@
+                       "'{1}' during DirRef::OpenDir",
+                       std::move(path), dfd_);
+    }
+    // Also check that the UID is the current effective UID. We don't currently
+    // verify the GID because it could come from the parent directory, so
+    // callers that need to should instead validate this themselves.
+    if (stat_result->uid() != geteuid()) {
+      // Model this as `EPERM`, which is a bit awkward, but should be fine.
+      return PathError(EPERM,
+                       "Setting UID when creating '{0}' relative to "
+                       "'{1}' during DirRef::OpenDir",
+                       std::move(path), dfd_);
+    }
+  }
+
+  return result;
+}
+
+auto MakeTmpDir() -> ErrorOr<RemovingDir, Error> {
+  std::filesystem::path tmpdir_path = "/tmp";
+  for (const char* tmpdir_env_name : {"TEST_TMPDIR", "TMPDIR"}) {
+    const char* tmpdir_env_cstr = getenv(tmpdir_env_name);
+    if (tmpdir_env_cstr == nullptr) {
+      continue;
+    }
+    std::filesystem::path tmpdir_env = std::string(tmpdir_env_cstr);
+    if (!tmpdir_env.is_absolute()) {
+      continue;
+    }
+    tmpdir_path = std::move(tmpdir_env);
+    break;
+  }
+
+  std::filesystem::path target = BuildData::BuildTarget.str();
+  std::string dir_name = target.filename().native();
+  dir_name += ".XXXXXX";
+
+  tmpdir_path /= dir_name;
+
+  std::string tmpdir_path_buffer = tmpdir_path.native();
+  char* result = mkdtemp(tmpdir_path_buffer.data());
+  if (result == nullptr) {
+    RawStringOstream os;
+    os << llvm::formatv("Calling mkdtemp on '{0}' failed: ",
+                        tmpdir_path.native());
+    PrintErrorNumber(os, errno);
+    return Error(os.TakeStr());
+  }
+  CARBON_CHECK(result == tmpdir_path_buffer.data());
+  tmpdir_path = std::move(tmpdir_path_buffer);
+
+  // Because `mkdtemp` doesn't return an open directory atomically, open the
+  // created directory and perform safety checks. We can be more strict here as
+  // there can't be correct racing creation of the *same* new temporary
+  // directory.
+  CARBON_ASSIGN_OR_RETURN(Dir tmp, Cwd().OpenDir(tmpdir_path));
+  // Make sure we try to remove the directory from here on out.
+  RemovingDir result_dir(std::move(tmp), tmpdir_path);
+
+  // It's a bit awkward to report `fstat` errors as `Error`s, but we
+  // don't have much choice. The stat failing here would be very weird.
+  CARBON_ASSIGN_OR_RETURN(FileStatus stat, result_dir.Stat());
+
+  // The permissions must be exactly 0700 for a temporary directory, and the UID
+  // should be ours.
+  if (stat.permissions() != 0700 && stat.uid() != geteuid()) {
+    return Error(
+        llvm::formatv("Found incorrect permissions or UID on tmpdir '{0}'",
+                      tmpdir_path.native())
+            .str());
+  }
+
+  // Last but not least, the directory must also be empty (other than `.` and
+  // `..`). Any files here would represent a security issue from injected
```

Comment:
Such symlinks could always appear later, no? I am not sure about framing this as a security issue. If it is, how does this prevent them from being added later?

Reply:
SGTM